src/typechecker.erl

```diff
@@ -1136,6 +1139,44 @@ new_type_var() ->
     I = erlang:unique_integer(),
     "_TyVar" ++ integer_to_list(I).
 
+-spec solve_bounds(#tenv{}, [type()]) -> #{ atom() := type() }.
+solve_bounds(TEnv, Cs) ->
+    Defs = [ {X, T} || {type, _, constraint, [{atom, _, is_subtype}, [{var, _, X}, T]]} <- Cs ],
+    Env  = lists:foldl(fun({X, T}, E) ->
+                maps:update_with(X, fun(Ts) -> [T | Ts] end, [T], E)
+            end, #{}, Defs),
+    DepG = maps:map(fun(_, T) -> maps:keys(free_vars(T)) end, Env),
+    SCCs = gradualizer_lib:top_sort(DepG),
+    solve_bounds(TEnv, Env, SCCs, #{}).
+
+solve_bounds(TEnv, Defs, [{acyclic, X} | SCCs], Acc) ->
+    #{ X := Tys } = Defs,
+    Tys1 = subst_ty(Acc, Tys),
+    %% Take intersection after substitution to get rid of type variables.
+    Ty1  = lists:foldl(fun(S, T) -> glb(S, T, TEnv) end, type(term), Tys1),
+    solve_bounds(TEnv, maps:remove(X, Defs), SCCs, Acc#{ X => Ty1 });
+solve_bounds(_, _, [{cyclic, Xs} | _], _) ->
+    throw({cyclic_dependencies, Xs});
```

Comment:
Here's a rather late comment, as I didn't realise this until now: If we have 'A :: B' and 'B :: A' then that amounts to A being equal to B. We could detect that but it's probably not worth it.

I think it'd be good to try to report an informative error message here, since it is fairly easy to accidentally write cyclic constraints by mistake for someone who uses constraints excessively. @UlfNorell, would you be willing to tackle that?

Reply:
Yes, I'm working on this. There are also some cases of cyclic dependencies in OTP. For instance
```erlang
-spec lists:flatten(DeepList) -> List when
  DeepList :: [term() | DeepList],
  List :: [term()].
```
The first step is to print a nice error, but we might indeed consider doing some work to solve the equations.

Reply:
#99